### PR TITLE
TST: fix read_stata doctest #42670

### LINKED ISSUE #42670: TST: Fix doctest in read_stata
# tl;dr

Fix the next doctest error:

```
___________________________________________________________________ [doctest] pandas.io.stata.read_stata ____________________________________________________________________
2044 Categorical variables read through an iterator may not have the same
2045 categories and dtype. This occurs when  a variable stored in a DTA
2046 file is associated to an incomplete set of value labels that only
2047 label a strict subset of the values.
2048 
2049 Examples
2050 --------
2051 Read a Stata dta file:
2052 
2053 >>> df = pd.read_stata('filename.dta')
UNEXPECTED EXCEPTION: FileNotFoundError(2, 'No such file or directory')
Traceback (most recent call last):
  File "/home/mgarcia/miniconda3/envs/pandas-dev/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pandas.io.stata.read_stata[0]>", line 1, in <module>
  File "/home/mgarcia/src/pandas/pandas/io/stata.py", line 1946, in read_stata
    reader = StataReader(
  File "/home/mgarcia/src/pandas/pandas/io/stata.py", line 1091, in __init__
    with get_handle(
  File "/home/mgarcia/src/pandas/pandas/io/common.py", line 710, in get_handle
    handle = open(handle, ioargs.mode)
FileNotFoundError: [Errno 2] No such file or directory: 'filename.dta'
/home/mgarcia/src/pandas/pandas/io/stata.py:2053: UnexpectedException

```

# Detailed instructions

Python allows to have example code in the documentation, like in:

```python
def add(num1, num2):
    """
    Computes the sum of the two numbers.

    Examples
    --------
    >>> add(2, 2)
    4
    """
    return num1 + num2
```

In pandas, we use this to document most elements. And there are tools, like pytest,
that can run the examples, and make sure everything is correct.

For historical reasons, we have many examples where the code fails to run, or the
actual output is different from the expected output. For example, check the next
incorrect examples:

```python
def add(num1, num2):
    """
    Computes the sum of the two numbers.

    Examples
    --------
    >>> add(2, 2)
    5

    >>> add(2, 2
    4

    >>> add(2, number)
    4

    ...
    """
    return num1 + num2
```

All them will fail for different reasons. To test the docstring of an object,
the next command can be run:

```
python -m pytest --doctest-modules pandas/core/frame.py::pandas.core.frame.DataFrame.info
```

Where `pandas/core/frame.py` is the file where the docstring is defined, and
`pandas.core.frame.DataFrame.info` is the object. A whole file can also be tested
by removing the `::` and the object from the command above.

In general, the errors in the examples can be fixed with things like:

- Fixing a typo (a missing comma, an mispelled variable name...)
- Adding an object that hasn't been defined (like, if `df` is used, but
  no sample dataset `df` has been first defined)
- Fixing the expected output, when it's wrong
- In exceptional cases, examples shouldn't run, since they can't work.
  For example, a function that connects to a private webservice. In
  such cases, we can add `# doctest: +SKIP` at the end of the lines
  that should not run

To be able to properly fix an example for the first time, the next steps
are needed:

- Install a pandas development environment in your computer. There are
  simplified instructions [in this page](https://python-sprints.github.io/pandas/guide/pandas_setup.html),
  and more detailed information in [pandas official contributing page](https://pandas.pydata.org/docs/development/index.html).
- Run the doctests for the object of interest (the one in this issue),
  and make sure the examples are still broken in the `master` branch of
  pandas
- Fix the file locally, and run the doctests again, to make sure the
  fix is working as expected
- Optionally have a look and make sure that the code in the examples
  follow PEP-8, and fix the style if it doesn't
- Commit your changes, push your branch to a fork, and open a pull
  request. Make sure you edit the line `Closes #XXXX` with the issue
  number you are addressing, so the issue is automatically closed,
  when the pull request is merged
- Make sure the continuous integration of your pull request finishes
  in green. If it doesn't, check if the problem is in your changes
  (sometimes things break in master for technical problems, and in
  that case you just need to wait for a core developer to fix the
  problem)
- Address any comment from the reviewers (just make changes locally,
  commit, and push to your branch, no need to open new pull requests)

### PR DESCRIPTION
**Pull request by KrishnaSai2020**
- [x] closes #42670
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] example docstring code fixed for read_strata

### CODE REVIEW

**Review by fangchenli — CHANGES_REQUESTED** [review]
We want them fixed rather than skipped.

**Comment by KrishnaSai2020** [comment]
Yes but the only solution I can see is creating a dummy .dta file for it to work since it needs a file to be read and I don't know where you would store that.

File: pandas/io/stata.py
Comment (KrishnaSai2020) [inline pandas/io/stata.py:165]: I've created this directory to store this dummy Stata file in pandas/core/io. ie the same base directory as this file but feel free to tell me to move the directory to a more appropriate place.

Comment (fangchenli) [reply pandas/io/stata.py:165]: You could construct a DataFrame, and do a `to_stata` in the doc section first. Then read it. Some of the other IO functions are like that.

See https://github.com/pandas-dev/pandas/blob/c86b129e517ff5489160d69b8eb399b68488b82e/pandas/io/pickle.py#L69

**Comment by KrishnaSai2020** [comment]
Made the requested changes @fangchenli.

**Comment by bashtage** [comment]
This is ruining the docstring purpose. It is not a good idea.  The entire point of reading with chunks to to parse a large data file.  IT should have > 10,000 roww to be illustrative.

**Comment by KrishnaSai2020** [comment]
@bashtage I see what you mean. I made the changes to the df's size please review the new changes.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:174]: size=(10,000,1) is not a legal Python expression.  You should go with

I would use

```
values = np.random.randint(0,10, size=(20_000,1), dtype="uint8")
df = pd.DataFrame(values, columns=["i"])
df.to_stata('filename.dta')
```

```
itr = pd.read_stata('filename.dta', chunksize=10000)
for chunk in itr:
    # Operate on a single chunk
    chunk.mean()
```

1. The type will ensure the file is always small. 
2. Setting the column will prevent a warning
3. 20k records means the iteration is non-trivial

**Comment by KrishnaSai2020** [comment]
@bashtage thanks for you help. I have now made the changes.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:174]: Need a space between `0,10`. Also needed between `20_000,1`.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:181]: Indentation is wrong here.  Need to align with 1 4 space indent from the `for chunk...` line.

**Review by bashtage — CHANGES_REQUESTED** [review]
Some formatting issues

**Comment by KrishnaSai2020** [comment]
I think I've fixed the formatting now. Please do let me know if there is anything else

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:180]: Too much indentation here.  # needs to come left 4.

**Review by bashtage — CHANGES_REQUESTED** [review]
Still a bit of formatting.

**Comment by KrishnaSai2020** [comment]
Think I've fixed them please review.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:174]: need spaces after the >>> throughout.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:165]: You must escape the `{` here since this is an f-string.  You need to use `{{` whenever you want a `{` in code.

**Review by bashtage — CHANGES_REQUESTED** [review]
Code is failing because of unescaped {

**Comment by bashtage** [comment]
Need to add `pandas/io/stata.py \ ` after this line so that the state doctests are run.

https://github.com/pandas-dev/pandas/blob/b8d750fb90c68d4af25af89442aba001483adffa/ci/code_checks.sh#L123

**Comment by KrishnaSai2020** [comment]
@bashtage I've added this dockest into the scripts and resolved the conflicts

**Comment by KrishnaSai2020** [comment]
This should now pass all checks.

**Comment by KrishnaSai2020** [comment]
From what I can tell everything works apart from the doctests in Style_render.py which is unrelated to this pr.

**Comment by KrishnaSai2020** [comment]
Could I please get an update as to what is going to happen with this pr, please? Is it going to get merged? or have I done something wrong which means it can get merged?

File: ci/code_checks.sh
Comment (bashtage) [inline ci/code_checks.sh:125]: Why did you add style_renderer?  You should remove this since it does not pass.

File: pandas/io/stata.py
Comment (bashtage) [inline pandas/io/stata.py:181]: Ideally the grammar would be corrected and this would read `... single chunk, e.g., chunk.mean()`

**Review by bashtage — CHANGES_REQUESTED** [review]
Need to remove new module that doesn't pass, and small grammar change

**Review by bashtage — APPROVED** [review]

**Comment by KrishnaSai2020** [comment]
@fangchenli please re-review

**Comment by KrishnaSai2020** [comment]
Could I get a second opinion on this PR, please? it's passed all checks and already has one approval

**Comment by bashtage** [comment]
@KrishnaSai2020 This LGMT.  I'll leave it to @jreback to make the final decision. You don't need to merge master in anymore unless there is a conflict.

**Comment by bashtage** [comment]
@KrishnaSai2020 Thanks.

**Pull request merged**
